File: Luogu/P_1216_USACO_1_5_IOI_1994_数字三角形_Number_Triangles.cpp
#include <bits/stdc++.h>
using namespace std;
const int INF = 0x3f3f3f3f;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int r; cin >> r;
    vector<vector<int>> rows(r, vector<int>(r));
    for (int i = 0; i < r; i++) {
        for (int j = 0; j <= i; j++) {
            cin >> rows[i][j];
        }
    }

    vector<vector<int>> memo(r, vector<int>(r, -INF));
    auto dfs = [&](auto &&dfs, int i, int j) -> int {
        if (i == r - 1) return rows[i][j];
        if (memo[i][j] != -INF) return memo[i][j];
        return memo[i][j] = rows[i][j] + max(dfs(dfs, i + 1, j), dfs(dfs, i + 1, j + 1));
    };
    cout << dfs(dfs, 0, 0) << endl;
    return 0;
}

File: Luogu/P_1216_USACO_1_5_IOI_1994_数字三角形_Number_Triangles.py
r = int(input())
rows = [list(map(int, input().split())) for _ in range(r)]

f = [[-1] * r for _ in range(r)]
f[0][0] = rows[0][0]
for i in range(1, r):
    for j in range(i + 1):
        f[i][j] = rows[i][j] + max(f[i - 1][j], f[i - 1][j - 1])
print(max(f[r - 1]))